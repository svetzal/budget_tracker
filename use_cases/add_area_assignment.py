from datetime import date
from typing import Optional

from entities import AreaAssignment, LEVEL_PROGRAM, LEVEL_EXECUTIVE
from use_cases.use_case import UseCase


class AddAreaAssignment(UseCase):
    def execute(self, support_area_code: str, person_code: str, funding_source_transit: int, start_date: str,
                end_date: str, level: Optional[str]):
        self.guard_support_area_must_exist(support_area_code)
        self.guard_person_must_exist(person_code)
        self.guard_funding_source_must_exist(funding_source_transit)
        if level not in [LEVEL_EXECUTIVE, LEVEL_PROGRAM]:
            raise ValueError(f"Level must be 'exec' or 'program', not {level}")
        area_assignment = AreaAssignment(
            support_area_code=support_area_code,
            person_code=person_code,
            funding_source_transit=funding_source_transit,
            level=level,
            start_date=date.fromisoformat(start_date),
            end_date=date.fromisoformat(end_date)
        )
        self.practice.area_assignments.append(area_assignment)
