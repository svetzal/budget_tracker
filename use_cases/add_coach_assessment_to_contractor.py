from entities import CoachAssessment
from use_cases.use_case import UseCase


class AddCoachAssessmentToContractor(UseCase):
    def execute(self, contractor_code: str, completed_on: str, leadership: float, technical: float, practice: float):
        coach = self.find_contractor(contractor_code)
        coach.assessments.append(
            CoachAssessment(
                completed_on=completed_on,
                leadership=leadership,
                technical=technical,
                practice=practice,
            )
        )
