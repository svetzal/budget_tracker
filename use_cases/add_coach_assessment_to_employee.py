from entities import CoachAssessment
from use_cases.use_case import UseCase


class AddCoachAssessmentToEmployee(UseCase):
    def execute(self, employee_code: str, completed_on: str, leadership: float, technical: float, practice: float):
        coach = self.find_employee(employee_code)
        coach.assessments.append(
            CoachAssessment(
                completed_on=completed_on,
                leadership=leadership,
                technical=technical,
                practice=practice,
            )
        )
