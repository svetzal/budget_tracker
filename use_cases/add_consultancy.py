from entities import Consultancy
from use_cases.use_case import UseCase


class AddConsultancy(UseCase):
    def execute(self, code: str, name: str, contract: str, contact_name: str, contact_phone: str, contact_email: str):
        self.guard_consultancy_duplicate(code)
        consultancy = Consultancy(
            code=code,
            name=name,
            contract=contract,
            contact_name=contact_name,
            contact_phone=contact_phone,
            contact_email=contact_email
        )
        self.practice.consultancies.append(consultancy)
