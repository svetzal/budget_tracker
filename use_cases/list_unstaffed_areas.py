import pandas as pd
from tabulate import tabulate

from use_cases.use_case import UseCase


class ListUnstaffedAreas(UseCase):
    def execute(self, start_date: str, end_date: str):
        period_start = pd.to_datetime(start_date)
        period_end = pd.to_datetime(end_date)

        report_date_range = pd.bdate_range(period_start, period_end, freq="C",
                                           holidays=self.practice.statutory_holiday_list)
        report_frame = pd.DataFrame(report_date_range, columns=["date"])
        assignments_frame = self.prepare_assignments_frame(report_date_range)
        report_frame = report_frame.merge(assignments_frame, how="outer", on="date")
        self.prepare_date_groupings(report_frame)
        lines = self.prepare_report_output(report_frame)
        return lines

    def prepare_report_output(self, report_frame):
        lines = []
        for a in self.practice.area_assignments:
            unstaffed = report_frame[report_frame[a.support_area_code] == False]
            report = unstaffed.groupby(f"{a.support_area_code}_group")['date'].agg(['min', 'max'])
            report = report.reset_index(drop=True)
            if report.count()[0] > 0:
                lines.append("")
                lines.append(f"{self.find_support_area(a.support_area_code).name} is Unstaffed")
                lines.append("")
                report['min'] = report['min'].dt.strftime('%Y-%m-%d')
                report['max'] = report['max'].dt.strftime('%Y-%m-%d')
                report.columns = ['Start Date', 'End Date']
                lines.append(tabulate(report, showindex=False, tablefmt='plain'))
        return lines

    def prepare_date_groupings(self, report_frame):
        for a in self.practice.area_assignments:
            report_frame[f"{a.support_area_code}_group"] = (report_frame[a.support_area_code].diff() != 0).cumsum()

    def prepare_assignments_frame(self, report_date_range):
        assignment_coverage = []
        columns = ['date'] + [a.support_area_code for a in self.practice.area_assignments]
        for date in report_date_range:
            row = [date]
            for a in self.practice.area_assignments:
                start = pd.to_datetime(a.start_date)
                end = pd.to_datetime(a.end_date)
                row.append(start <= date and end >= date)
            assignment_coverage.append(row)
        assignments_frame = pd.DataFrame(assignment_coverage, columns=columns)
        return assignments_frame


