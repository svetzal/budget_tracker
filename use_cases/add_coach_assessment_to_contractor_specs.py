from entities import CoachingPracticeFinance
from use_cases.add_coach_assessment_to_contractor import AddCoachAssessmentToContractor
from use_cases.test_data_generators import FIRST_CONTRACTOR_CODE
from verifiers import verify_model


def test_can_add_coach_assessment_to_contractor(staffed_practice: CoachingPracticeFinance):
    AddCoachAssessmentToContractor(staffed_practice).execute(
        contractor_code=FIRST_CONTRACTOR_CODE,
        completed_on="2023-01-01",
        leadership=0.5,
        technical=0.5,
        practice=1.0,
    )

    verify_model(staffed_practice.contractors)
