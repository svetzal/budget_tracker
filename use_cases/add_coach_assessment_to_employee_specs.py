from entities import CoachingPracticeFinance
from use_cases.add_coach_assessment_to_employee import AddCoachAssessmentToEmployee
from use_cases.test_data_generators import FIRST_EMPLOYEE_CODE
from verifiers import verify_model


def test_can_add_coach_assessment_to_employee(staffed_practice: CoachingPracticeFinance):
    AddCoachAssessmentToEmployee(staffed_practice).execute(
        employee_code=FIRST_EMPLOYEE_CODE,
        completed_on="2023-01-01",
        leadership=0.5,
        technical=0.5,
        practice=1.0,
    )

    verify_model(staffed_practice.employees)
