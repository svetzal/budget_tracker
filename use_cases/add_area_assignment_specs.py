from _decimal import Decimal

from entities import CoachingPracticeFinance
from model_verifier import verify_model
from use_cases.add_area_assignment import AddAreaAssignment
from use_cases.add_consultancy import AddConsultancy
from use_cases.add_contractor import AddContractor
from use_cases.add_funding_source import AddFundingSource
from use_cases.add_support_area import AddSupportArea


def test_add_area_assignment():
    practice = CoachingPracticeFinance()

    AddSupportArea(practice).execute(
        code="1",
        name="Area 1"
    )

    AddFundingSource(practice).execute(
        transit=1,
        name="Funding Source 1",
        total=Decimal(100000),
        start_date="2020-01-01",
        end_date="2020-12-31",
    )

    AddConsultancy(practice).execute(
        code="1",
        name="Consultancy 1",
        contract="CTR00123",
        contact_name="Someone",
        contact_phone="647-555-1212",
        contact_email="someone@somewhere.com",
    )

    AddContractor(practice).execute(
        code="1",
        name="Contractor 1",
        consultancy_code="1",
        phone_number="647-555-1212",
        email="someone@somewhere.com",
        start_date="2020-01-01",
    )

    AddAreaAssignment(practice).execute(
        support_area_code="1",
        contractor_code="1",
        funding_source_transit=1,
        start_date="2020-01-01",
        end_date="2020-12-31",
    )

    verify_model(practice.area_assignments)