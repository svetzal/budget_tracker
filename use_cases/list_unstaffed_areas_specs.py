from entities import CoachingPracticeFinance
from use_cases.list_unstaffed_areas import ListUnstaffedAreas
from verifiers import verify_report


def test_list_unstaffed_areas(operational_practice: CoachingPracticeFinance):
    output = ListUnstaffedAreas(operational_practice).execute(start_date="2022-11-01", end_date="2023-10-31")

    verify_report(output)