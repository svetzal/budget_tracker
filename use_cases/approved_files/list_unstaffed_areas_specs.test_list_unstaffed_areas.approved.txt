
Area 1 is Unstaffed

2022-11-01  2022-12-30
